src/main/scala/mesosphere/marathon/api/TaskKiller.scala

```diff
@@ -100,8 +103,15 @@ class TaskKiller @Inject() (
       toKill = appTasks
     )
 
-    appTasks.keys.find(id => !instanceTracker.hasSpecInstancesSync(id))
-      .map(id => Future.failed(PathNotFoundException(id)))
-      .getOrElse(killTasks)
+    val idHasSpec = Future.sequence(appTasks.keys.map(id => async {
```

Comment:
Instead of calling `instanceTracker.hasSpecInstances` for every appId, you could get all instances by spec once:
```scala
val allInstances = await(instanceTracker.instancesBySpec())
val (existing, nonExisting) = appTasks.keys.partition(allInstances.hasSpecInstance)
if (nonExisting.nonEmpty) ...
```